Graphite: Migrate to React (part 1: move state to redux)

ifrost opened the pull request:
This is first PR for Graphite to React migration. I'm trying to replace the engine while driving the car here. More details in #36481.

First step is to move all logic from Angular controller to redux state/reduces. They will be used by React components later. No new components, no changes to the UI or the behavior. The logic is almost identical - I just moved functions to different places. I kept empty functions in GraphiteQueryCtrl with some comments for easier diffing.

All functions from GraphiteQueryCtrl have been moved to separate files:
- state/actions - with all available UI actions
- state/providers - hooks providing data for autocomplete
- state/store - main store and reducers
- state/helpers - common logic used by reducers and providers

How to test it?

Graphite is a complex beast. You can use [manual test scenarios ](https://gist.github.com/ifrost/e8faa2976caf610e9e03ec815ce83d82) or simply play with it compering the behavior with main/play.

ifrost committed:
Add UMLs

ifrost committed:
Add rendered diagrams

ifrost committed:
Move QueryCtrl to flux

ifrost committed:
Remove redundant param in the reducer

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

# Conflicts:
#	public/app/plugins/datasource/graphite/partials/query.editor.html
#	public/app/plugins/datasource/graphite/query_ctrl.ts
#	public/app/plugins/datasource/graphite/types.ts

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Use named imports for lodash and fix typing for GraphiteTagOperator

ifrost committed:
Add missing async/await

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Extract providers to a separate file

ifrost committed:
Clean up async await

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Rename controller functions back to main

ifrost committed:
Simplify creating actions

ifrost committed:
Re-order controller functions

ifrost committed:
Separate helpers from actions

ifrost committed:
Rename vars

ifrost committed:
Simplify helpers

ifrost committed:
Move controller methods to state reducers

ifrost committed:
Remove docs (they are added in design doc)

ifrost committed:
Move actions.ts to state folder

ifrost committed:
Add docs

ifrost committed:
Add old methods stubs for easier review

ifrost committed:
Check how state dependencies will be mapped

ifrost committed:
Rename state to store

ifrost committed:
Rename state to store

ifrost committed:
Rewrite spec tests for Graphite Query Controller

ifrost committed:
Update docs

torkelo commented:
Nice! 
did you take a look at the Elastic query editor? think it uses useReducer instead of redux ,  written by @Elfo404 . maybe there are some patterns there we can reuse. not that there is much different with functional components now with useReducer  vs useDispatch

ifrost committed:
Update docs

ifrost commented:
> Nice!
> did you take a look at the Elastic query editor? think it uses useReducer instead of redux , written by @Elfo404 . maybe there are some patterns there we can reuse. not that there is much different with functional components now with useReducer vs useDispatch

Yes, I think it makes total sense to make it very similar to Elastic and maybe share some code for creating ContextProviders. I'll try to tackle it when I get to the point when I can remove Angular controller. As an intermediate step I thought I could simply pass down `state` / `dispatch` to make it easier to refactor all component bottom-up and add the root React component at the end.

Elfo404 commented on public/app/plugins/datasource/graphite/state/providers.ts:
```diff
@@ -0,0 +1,190 @@
+  remove(altSegments, (s) => s.value === '_tagged');
+}
+
+function mapToDropdownOptions(results: any[]) {
```
```suggestion
function mapToDropdownOptions(results: string[]) {
```

ifrost committed:
Update public/app/plugins/datasource/graphite/state/helpers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/helpers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/helpers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/providers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/providers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/providers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/providers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Update public/app/plugins/datasource/graphite/state/providers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost commented on public/app/plugins/datasource/graphite/state/providers.ts:
```diff
@@ -0,0 +1,190 @@
+export async function getAltSegments(
+  state: GraphiteQueryEditorState,
+  index: number,
+  prefix: string
+): Promise<GraphiteSegment[]> {
+  let query = prefix && prefix.length > 0 ? '*' + prefix + '*' : '*';
```
Prefix might be an empty string but should always be defined. I'll pick the second suggestion 👍

ifrost committed:
Update public/app/plugins/datasource/graphite/state/providers.ts

Co-authored-by: Giordano Ricci <me@giordanoricci.com>

ifrost committed:
Add more type definitions

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Merge branch 'ifrost/graphite-on-react-1' of https://github.com/grafana/grafana into ifrost/graphite-on-react-1

ifrost commented:
@Elfo404 - I've addressed your suggestions 👍 I've also added more types. It's surfaced a bug with missing `target` property when checking when to refresh the panel [here](https://github.com/grafana/grafana/pull/36484/commits/a952d47c8abe08942ae0b388e959384ff1e244a9#diff-403fc986757091098706d2088928dae68ebe907e5fc50b0b0b54d3f5fdf1aa9aR183) 😎

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

aocenas reviewed (commented):
Nice, I did not test it thoroughly but like where this is going. Also like explanatory comments in the code makes following the changes much easier.

ifrost committed:
Load function definitions before parsing the target on initial load

ifrost commented:
a3f13ee: small change to fix loading function definitions. They should be loaded before the model is created to correctly recognize already defined functions. 

query_ctrl.tsx (main)
```typescript
this.datasource.waitForFuncDefsLoaded().then(() => {                              // load functions
  this.queryModel = new GraphiteQuery(this.datasource, this.target, templateSrv); // create model
  this.buildSegments(false);                                                      // build ui
});
```

store.tsx (branch)
```typescript
  if (actions.init.match(action)) {
    await deps.datasource.waitForFuncDefsLoaded();                                    // load functions
    state = {
      ...
      queryModel: new GraphiteQuery(deps.datasource, deps.target, getTemplateSrv()),  // create model
    };
    await buildSegments(state, false);                                                // build ui
  }
```

ifrost committed:
Change targetChanged to updateQuery to avoid mutating state directly

It's also needed for extra refresh/runQuery execution as handleTargetChanged doesn't handle changing the raw query

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Fix updating query after adding a function

ifrost committed:
Simplify updating function params

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1

Elfo404 reviewed (approved):
Just a small nit (aside from the async things discussed offline), but looks great and seems to work as expected! 🎉 

I really like how you migrated the state before during UI work, great stuff in here! 🔥

aocenas reviewed (approved):
Tried it a bit and seems to work the same as previous version.

There is a bug that seems to be in both versions where adding a tag and removing it then prevents adding a metric:

https://user-images.githubusercontent.com/1014802/126306253-f6594f47-de51-4ba7-80ca-871e51efe3b6.mp4

I guess we can later fix it separatelly

ifrost committed:
Remove redundant awaits

ifrost committed:
Use redux Action

ifrost commented on public/app/plugins/datasource/graphite/state/store.ts:
```diff
@@ -0,0 +1,177 @@
+  metricAutoCompleteErrorShown: boolean;
+};
+
+type Action = {
+  type: string;
+  payload: any;
+};
```
> Tried it a bit and seems to work the same as previous version.
> 
> There is a bug that seems to be in both versions where adding a tag and removing it then prevents adding a metric:
> 
>  Peek.2021-07-20.12-07.mp4 
> I guess we can later fix it separatelly

Yep - it's the bug in the current implementation. I've added a note to test cases scenarios. I'll try to fix different bugs once all components are migrated.

ifrost replied:
Absolutely - update with feed0cd

ifrost commented:
> Tried it a bit and seems to work the same as previous version.
> 
> There is a bug that seems to be in both versions where adding a tag and removing it then prevents adding a metric:
> 
>  Peek.2021-07-20.12-07.mp4 
> I guess we can later fix it separatelly

Yep - it's the bug in the current implementation. I've added a note to test cases scenarios. I'll try to fix different bugs once all components are migrated.

ifrost committed:
Use more specific type for GraphiteTag

ifrost commented on public/app/plugins/datasource/graphite/state/providers.ts:
```diff
@@ -0,0 +1,190 @@
+
+export async function getTagValues(
+  state: GraphiteQueryEditorState,
+  tag: { key: any },
```
Good spot - c18ff18

ifrost committed:
Merge branch 'main' into ifrost/graphite-on-react-1
